streams/test-utils/src/main/java/org/apache/kafka/streams/test/TopologyTestDriver.java

```diff
@@ -0,0 +1,593 @@
+ */
+@InterfaceStability.Evolving
+public class TopologyTestDriver {
+
+    private final Time mockTime;
+    private final InternalTopologyBuilder internalTopologyBuilder;
+
+    private final static int PARTITION_ID = 0;
+    private final static TaskId TASK_ID = new TaskId(0, PARTITION_ID);
+    private StreamTask task;
+    private GlobalStateUpdateTask globalStateTask;
+
+    private final ProcessorTopology processorTopology;
+    private final MockProducer<byte[], byte[]> producer;
+
+    private final Set<String> internalTopics = new HashSet<>();
+    private final Map<String, TopicPartition> partitionsByTopic = new HashMap<>();
+    private final Map<String, TopicPartition> globalPartitionsByTopic = new HashMap<>();
+    private final Map<TopicPartition, AtomicLong> offsetsByTopicPartition = new HashMap<>();
+
+    private final Map<String, Queue<ProducerRecord<byte[], byte[]>>> outputRecordsByTopic = new HashMap<>();
+
+    /**
+     * Create a new test diver instance.
+     * Initialized the internally mocked wall-clock time with {@link System#currentTimeMillis() current system time}.
+     *
+     * @param topology the topology to be tested
+     * @param config the configuration for the topology
+     */
+    public TopologyTestDriver(final Topology topology,
+                              final Properties config) {
+        this(topology, config, System.currentTimeMillis());
+    }
+    /**
+     * Create a new test diver instance.
+     *
+     * @param topology the topology to be tested
+     * @param config the configuration for the topology
+     * @param initialWallClockTimeMs the initial value of internally mocked wall-clock time
+     */
+    public TopologyTestDriver(final Topology topology,
+                              final Properties config,
+                              final long initialWallClockTimeMs) {
+        final StreamsConfig streamsConfig = new StreamsConfig(config);
+        mockTime = new MockTime(initialWallClockTimeMs);
+
+        internalTopologyBuilder = InternalTopologyBuilderAccessor.getInternalTopologyBuilder(topology);
+        internalTopologyBuilder.setApplicationId(streamsConfig.getString(StreamsConfig.APPLICATION_ID_CONFIG));
+
+        processorTopology = internalTopologyBuilder.build(null);
+        final ProcessorTopology globalTopology  = internalTopologyBuilder.buildGlobalStateTopology();
+
+        final Serializer<byte[]> bytesSerializer = new ByteArraySerializer();
+        producer = new MockProducer<byte[], byte[]>(true, bytesSerializer, bytesSerializer) {
+            @Override
+            public List<PartitionInfo> partitionsFor(final String topic) {
+                return Collections.singletonList(new PartitionInfo(topic, PARTITION_ID, null, null, null));
+            }
+        };
+
+        final MockConsumer<byte[], byte[]> consumer = new MockConsumer<>(OffsetResetStrategy.EARLIEST);
+        final StateDirectory stateDirectory = new StateDirectory(streamsConfig, mockTime);
```

Comment:
If user do not modify the state directory path in their unit tests, does that mean two unit tests would conflict on the state directory path then? Should we just use a random directory than enforcing themselves to remember setting to a different path each time?